src/hotspot/os/aix/os_perf_aix.cpp

```diff
@@ -573,267 +433,85 @@ int CPUPerformanceInterface::context_switch_rate(double* rate) const {
 }
 
-char* SystemProcessInterface::SystemProcesses::ProcessIterator::allocate_string(const char* str) const {
+char* SystemProcessInterface::SystemProcesses::allocate_string(const char* str) const {
   if (str != NULL) {
     return os::strdup_check_oom(str, mtInternal);
   }
   return NULL;
 }
 
-int SystemProcessInterface::SystemProcesses::ProcessIterator::current(SystemProcess* process_info) {
-  if (!is_valid()) {
-    return OS_ERR;
-  }
+int SystemProcessInterface::SystemProcesses::system_processes(SystemProcess** system_processes, int* nprocs) const {
+  ResourceMark rm;
+  perfstat_process_t* proc_stats;
+  SystemProcess* head;
+  perfstat_id_t name_holder;
+  int records_allocated = 0;
 
-  process_info->set_pid(atoi(_entry->d_name));
+  assert(nprocs != NULL, "system_processes counter pointers is NULL!");
 
-  get_exe_name();
-  process_info->set_name(allocate_string(_exeName));
-
-  if (get_exe_path() != NULL) {
-     process_info->set_path(allocate_string(_exePath));
-  }
+  head = NULL;
+  *nprocs = 0;
+  strncpy(name_holder.name, "", IDENTIFIER_LENGTH);
 
-  char* cmdline = NULL;
-  cmdline = get_cmdline();
-  if (cmdline != NULL) {
-    process_info->set_command_line(allocate_string(cmdline));
-    FREE_C_HEAP_ARRAY(char, cmdline);
-  }
-
-  return OS_OK;
-}
-
-int SystemProcessInterface::SystemProcesses::ProcessIterator::next_process() {
-  if (!is_valid()) {
+  // calling perfstat_<subsystem>(NULL, NULL, _, 0) returns number of available records
+  *nprocs = libperfstat::perfstat_process(NULL, NULL, sizeof(perfstat_process_t), 0);
+  if(*nprocs < 1) {
+    // expect at least 1 process
     return OS_ERR;
   }
 
-  do {
-    _entry = os::readdir(_dir);
-    if (_entry == NULL) {
-      // Error or reached end.  Could use errno to distinguish those cases.
-      _valid = false;
-      return OS_ERR;
-    }
-  } while(!is_valid_entry(_entry));
-
-  _valid = true;
-  return OS_OK;
-}
-
-SystemProcessInterface::SystemProcesses::ProcessIterator::ProcessIterator() {
-  _dir = NULL;
-  _entry = NULL;
-  _valid = false;
-}
+  records_allocated = *nprocs;
+  proc_stats = NEW_RESOURCE_ARRAY(perfstat_process_t, records_allocated);
 
-bool SystemProcessInterface::SystemProcesses::ProcessIterator::initialize() {
-  // Not yet implemented.
-  return false;
-}
+  // populate stats && set the actual number of procs that have been populated
+  // should never be higher than requested, but may be lower due to process death
+  *nprocs = libperfstat::perfstat_process(&name_holder, proc_stats, sizeof(perfstat_process_t), records_allocated);
 
-SystemProcessInterface::SystemProcesses::ProcessIterator::~ProcessIterator() {
-  if (_dir != NULL) {
-    os::closedir(_dir);
-  }
-}
+  for (int n = 0; n < *nprocs; n++) {
+    psinfo_t psinfo;
+    char* name     = NEW_C_HEAP_ARRAY(char, IDENTIFIER_LENGTH, mtInternal);
+    char* exe_name = NEW_C_HEAP_ARRAY(char, PRFNSZ, mtInternal);
+    char* cmd_line = NEW_C_HEAP_ARRAY(char, PRARGSZ, mtInternal);
```

Comment:
So the contract with SystemProcess is that its ctor arguments need to be C-heap allocated, and releases them itself? ... okay. May be worthwhile to clean up at some point, or at least comment.

Reply:
Yeah, this was a bit of a gotcha during development (especially since NetworkInterface is different); I agree it's worth documenting.